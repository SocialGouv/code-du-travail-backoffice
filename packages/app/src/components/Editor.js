import styled from "@emotion/styled";
import { omit } from "ramda";
import React from "react";
import { Flex } from "rebass";

import markdown from "../libs/markdown";

const TextButton = styled.button`
  color: #444444;
  font-size: 0.875rem;
  width: auto !important;
`;
const TextButtonIcon = styled.img`
  height: 0.75rem;
  margin: 2px 3px 0;
  vertical-align: -1px;
  width: 0.75rem;
`;

/**
 * React wrapper for Quill plugin converting the HTML source into Markdown.
 *
 * @description
 * The `defaultValueDelta` prop is expecting a Quill delta.
 *
 * @see https://quilljs.com/docs/delta/
 */
export default class Editor extends React.Component {
  // Quill can't be tested sicnce it's not React-aware.
  // It will be tested via e2e tests.
  /* istanbul ignore next */
  componentDidMount() {
    const quillOptions = {
      format: ["blockquote", "bold", "header", "italic", "link", "list", "underline"],
      modules: {
        toolbar: {
          container: this.$toolbar,
          handlers: {
            redo: () => this.quill.history.redo(),
            undo: () => this.quill.history.undo()
          }
        }
      },

      theme: "snow"
    };

    // We have to load Quill here because of the global `document`
    // variable check done by the original library.
    const Quill = require("quill");
    this.quill = new Quill(this.$editor, quillOptions);
    this.quill.on("text-change", this.onChange.bind(this));

    this.quill.focus();
  }

  // Quill can't be tested in unit tests sicnce it's not React-aware.
  // It will be tested via e2e tests.
  /* istanbul ignore next */
  onChange() {
    if (this.props.onChange === undefined) return;

    const markdownContent = markdown.fromHtml(this.$editor.innerHTML);
    this.props.onChange(markdownContent);
  }

  getHtmlSource() {
    const htmlContent = markdown.toHtml(this.props.defaultValue);

    return { __html: htmlContent };
  }

  render() {
    const props = omit(
      ["defaultValue", "onChange", "onReferencesClicked", "onTagsClicked"],
      this.props
    );

    return (
      <Flex flexDirection="column" ref={el => (this.$container = el)} {...props}>
        <div className="ql-toolbar ql-snow" ref={el => (this.$toolbar = el)}>
          <span className="ql-formats">
            <TextButton
              className="ql-undo"
              title="Bouton annulant la dernière modification effectuée"
              type="button"
            >
              <TextButtonIcon src="/static/images/undo.svg" />
            </TextButton>
            <TextButton
              className="ql-redo"
              title="Bouton réappliquant la dernière modification annulée"
              type="button"
            >
              <TextButtonIcon src="/static/images/redo.svg" />
            </TextButton>
          </span>
          <span className="ql-formats">
            <button
              className="ql-bold"
              title="Bouton formattant le texte sélectionné en gras"
              type="button"
            />
            <button
              className="ql-italic"
              title="Bouton formattant le texte sélectionné en italique"
              type="button"
            />
            <button
              className="ql-underline"
              title="Bouton formattant le texte sélectionné en souligné"
              type="button"
            />
          </span>
          <span className="ql-formats">
            <button
              className="ql-link"
              title="Bouton créant un lien à partir du text sélectionné"
              type="button"
            />
            <button
              className="ql-blockquote"
              title={`Bouton formattant le paragraphe sélectionné au format citation`}
              type="button"
            />
          </span>
          <span className="ql-formats">
            <button
              className="ql-list"
              title={`Bouton formattant le paragraphe sélectionné au format liste ordonnée`}
              value="ordered"
            />
            <button
              className="ql-list"
              title={`Bouton formattant le paragraphe sélectionné au format liste numérotée`}
              value="bullet"
            />
          </span>
          <span className="ql-formats">
            <select className="ql-header" defaultValue="">
              <option value="">Normal</option>
              <option value="2">Titre</option>
              <option value="3">Sous-titre</option>
            </select>
          </span>
        </div>
        <div
          dangerouslySetInnerHTML={this.getHtmlSource()}
          ref={el => (this.$editor = el)}
          width={1}
        />
      </Flex>
    );
  }
}
