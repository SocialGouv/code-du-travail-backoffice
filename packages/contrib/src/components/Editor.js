import { omit } from "ramda";
import React from "react";
import { Flex } from "rebass";
import styled from "styled-components";

import markdown from "../lib/markdown";

const TextButton = styled.button`
  color: #444444;
  font-size: 0.875rem;
  width: auto !important;
`;

/**
 * React wrapper for Quill plugin converting the HTML source into Markdown.
 *
 * @description
 * The `defaultValueDelta` prop is expecting a Quill delta.
 *
 * @see https://quilljs.com/docs/delta/
 */
export default class Editor extends React.Component {
  componentDidMount() {
    const quillOptions = {
      theme: "snow",
      format: [
        "blockquote",
        "bold",
        "header",
        "italic",
        "link",
        "list",
        "underline"
      ],

      modules: {
        toolbar: {
          container: this.$toolbar,
          handlers: {
            charter: () =>
              window.open("/static/docs/Charte-Redactionnelle-v1.0.pdf"),
            redo: () => this.quill.history.redo(),
            tags: this.props.onTagsClicked,
            undo: () => this.quill.history.undo()
          }
        }
      }
    };

    // We have to load Quill here because of the global `document`
    // variable check done by the original library.
    const Quill = require("quill");
    this.quill = new Quill(this.$editor, quillOptions);
    this.quill.on("text-change", this.onChange.bind(this));

    this.quill.focus();
  }

  onChange() {
    if (this.props.onChange === undefined) return;

    const markdownContent = markdown.fromHtml(this.$editor.innerHTML);
    this.props.onChange(markdownContent);
  }

  getHtmlSource() {
    const htmlContent = markdown.toHtml(this.props.defaultValue);

    return { __html: htmlContent };
  }

  render() {
    const props = omit(
      ["defaultValue", "onChange", "onTagsClicked"],
      this.props
    );

    return (
      <Flex
        flexDirection="column"
        ref={el => (this.$container = el)}
        {...props}
      >
        <div className="ql-toolbar ql-snow" ref={el => (this.$toolbar = el)}>
          <span className="ql-formats">
            <TextButton type="button" className="ql-undo">
              Undo
            </TextButton>
            <TextButton type="button" className="ql-redo">
              Redo
            </TextButton>
          </span>
          <span className="ql-formats">
            <button type="button" className="ql-bold" />
            <button type="button" className="ql-italic" />
            <button type="button" className="ql-underline" />
          </span>
          <span className="ql-formats">
            <button type="button" className="ql-link" />
            <button type="button" className="ql-blockquote" />
          </span>
          <span className="ql-formats">
            <select className="ql-header" defaultValue="">
              <option value="">Normal</option>
              <option value="2">Titre</option>
              <option value="3">Sous-titre</option>
            </select>
          </span>
          <span className="ql-formats">
            <TextButton type="button" className="ql-charter">
              Charte
            </TextButton>
          </span>
          <span className="ql-formats">
            <TextButton type="button" className="ql-tags">
              Tags
            </TextButton>
          </span>
        </div>
        <div
          dangerouslySetInnerHTML={this.getHtmlSource()}
          ref={el => (this.$editor = el)}
          width={1}
        />
      </Flex>
    );
  }
}
