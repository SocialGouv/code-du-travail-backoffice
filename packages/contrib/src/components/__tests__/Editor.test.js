import React from "react";
import { render } from "@testing-library/react";

import Editor from "../Editor";

describe("<Editor />", () => {
  const props = {
    defaultValue: `
## Guerre et Paix

### Citation

> Il m'est arrivé de sentir que tout allait bien pour moi, que tout le<br>
> monde était gai, et aussitôt l'idée me traversait l'esprit qu'il ne se<br>
> passerait plus rien et que tout était absurde.

*Léon Tolstoï*

Ceci est <u>important</u>.

Ceci est très **important**.
    `,
    label: "A label",
    onChange: jest.fn(),
    onTagsClicked: jest.fn()
  };

  const γ = render(<Editor {...props} />);

  it("should match snapshot", () => {
    // We need to disable this attribute since it's automatically generated and
    // we can't have any control over it:
    // eslint-disable-next-line prettier/prettier
    γ.container.innerHTML = γ.container.innerHTML
      .replace(/ medium-editor-index="[^"]{36}"/, "");

    expect(γ.container).toMatchSnapshot();
  });
});
