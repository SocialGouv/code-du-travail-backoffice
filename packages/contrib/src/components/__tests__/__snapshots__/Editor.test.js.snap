// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Editor /> should match snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa sc-bwzfXH eVPJMi"
    label="A label"
  >
    <div
      class="ql-toolbar ql-snow"
    >
      <span
        class="ql-formats"
      >
        <button
          class="ql-undo Editor__TextButton-sc-1s8484a-0 eNdwXj"
          type="button"
        >
          Undo
        </button>
        <button
          class="ql-redo Editor__TextButton-sc-1s8484a-0 eNdwXj"
          type="button"
        >
          Redo
        </button>
      </span>
      <span
        class="ql-formats"
      >
        <button
          class="ql-bold"
          type="button"
        >
          <svg
            viewBox="0 0 18 18"
          >
             
            <path
              class="ql-stroke"
              d="M5,4H9.5A2.5,2.5,0,0,1,12,6.5v0A2.5,2.5,0,0,1,9.5,9H5A0,0,0,0,1,5,9V4A0,0,0,0,1,5,4Z"
            />
             
            <path
              class="ql-stroke"
              d="M5,9h5.5A2.5,2.5,0,0,1,13,11.5v0A2.5,2.5,0,0,1,10.5,14H5a0,0,0,0,1,0,0V9A0,0,0,0,1,5,9Z"
            />
             
          </svg>
        </button>
        <button
          class="ql-italic"
          type="button"
        >
          <svg
            viewBox="0 0 18 18"
          >
             
            <line
              class="ql-stroke"
              x1="7"
              x2="13"
              y1="4"
              y2="4"
            />
             
            <line
              class="ql-stroke"
              x1="5"
              x2="11"
              y1="14"
              y2="14"
            />
             
            <line
              class="ql-stroke"
              x1="8"
              x2="10"
              y1="14"
              y2="4"
            />
             
          </svg>
        </button>
        <button
          class="ql-underline"
          type="button"
        >
          <svg
            viewBox="0 0 18 18"
          >
             
            <path
              class="ql-stroke"
              d="M5,3V9a4.012,4.012,0,0,0,4,4H9a4.012,4.012,0,0,0,4-4V3"
            />
             
            <rect
              class="ql-fill"
              height="1"
              rx="0.5"
              ry="0.5"
              width="12"
              x="3"
              y="15"
            />
             
          </svg>
        </button>
      </span>
      <span
        class="ql-formats"
      >
        <button
          class="ql-link"
          type="button"
        >
          <svg
            viewBox="0 0 18 18"
          >
             
            <line
              class="ql-stroke"
              x1="7"
              x2="11"
              y1="7"
              y2="11"
            />
             
            <path
              class="ql-even ql-stroke"
              d="M8.9,4.577a3.476,3.476,0,0,1,.36,4.679A3.476,3.476,0,0,1,4.577,8.9C3.185,7.5,2.035,6.4,4.217,4.217S7.5,3.185,8.9,4.577Z"
            />
             
            <path
              class="ql-even ql-stroke"
              d="M13.423,9.1a3.476,3.476,0,0,0-4.679-.36,3.476,3.476,0,0,0,.36,4.679c1.392,1.392,2.5,2.542,4.679.36S14.815,10.5,13.423,9.1Z"
            />
             
          </svg>
        </button>
        <button
          class="ql-blockquote"
          type="button"
        >
          <svg
            viewBox="0 0 18 18"
          >
             
            <rect
              class="ql-fill ql-stroke"
              height="3"
              width="3"
              x="4"
              y="5"
            />
             
            <rect
              class="ql-fill ql-stroke"
              height="3"
              width="3"
              x="11"
              y="5"
            />
             
            <path
              class="ql-even ql-fill ql-stroke"
              d="M7,8c0,4.031-3,5-3,5"
            />
             
            <path
              class="ql-even ql-fill ql-stroke"
              d="M14,8c0,4.031-3,5-3,5"
            />
             
          </svg>
        </button>
      </span>
      <span
        class="ql-formats"
      >
        <span
          class="ql-header ql-picker"
        >
          <span
            aria-controls="ql-picker-options-0"
            aria-expanded="false"
            class="ql-picker-label"
            data-label="Normal"
            data-value=""
            role="button"
            tabindex="0"
          >
            <svg
              viewBox="0 0 18 18"
            >
               
              <polygon
                class="ql-stroke"
                points="7 11 9 13 11 11 7 11"
              />
               
              <polygon
                class="ql-stroke"
                points="7 7 9 5 11 7 7 7"
              />
               
            </svg>
          </span>
          <span
            aria-hidden="true"
            class="ql-picker-options"
            id="ql-picker-options-0"
            tabindex="-1"
          >
            <span
              class="ql-picker-item ql-selected"
              data-label="Normal"
              data-value=""
              role="button"
              tabindex="0"
            />
            <span
              class="ql-picker-item"
              data-label="Titre"
              data-value="2"
              role="button"
              tabindex="0"
            />
            <span
              class="ql-picker-item"
              data-label="Sous-titre"
              data-value="3"
              role="button"
              tabindex="0"
            />
          </span>
        </span>
        <select
          class="ql-header"
          style="display: none;"
        >
          <option
            selected=""
            value=""
          >
            Normal
          </option>
          <option
            value="2"
          >
            Titre
          </option>
          <option
            value="3"
          >
            Sous-titre
          </option>
        </select>
      </span>
      <span
        class="ql-formats"
      >
        <button
          class="ql-charter Editor__TextButton-sc-1s8484a-0 eNdwXj"
          type="button"
        >
          Charte
        </button>
      </span>
      <span
        class="ql-formats"
      >
        <button
          class="ql-tags Editor__TextButton-sc-1s8484a-0 eNdwXj"
          type="button"
        >
          Tags
        </button>
      </span>
    </div>
    <div
      class="ql-container ql-snow"
      width="1"
    >
      <div
        class="ql-editor"
        contenteditable="true"
        data-gramm="false"
      >
        <h2>
          Guerre et Paix
        </h2>
        <h3>
          Citation
        </h3>
        <blockquote>
          Il m'est arrivé de sentir que tout allait bien pour moi, que tout le
        </blockquote>
        <blockquote>
           monde était gai, et aussitôt l'idée me traversait l'esprit qu'il ne se
        </blockquote>
        <blockquote>
           passerait plus rien et que tout était absurde.
        </blockquote>
        <p>
          <em>
            Léon Tolstoï
          </em>
        </p>
        <p>
          Ceci est 
          <u>
            important
          </u>
          .
        </p>
        <p>
          Ceci est très 
          <strong>
            important
          </strong>
          .
        </p>
      </div>
      <div
        class="ql-clipboard"
        contenteditable="true"
        tabindex="-1"
      />
      <div
        class="ql-tooltip ql-hidden"
      >
        <a
          class="ql-preview"
          href="about:blank"
          target="_blank"
        />
        <input
          data-formula="e=mc^2"
          data-link="https://quilljs.com"
          data-video="Embed URL"
          type="text"
        />
        <a
          class="ql-action"
        />
        <a
          class="ql-remove"
        />
      </div>
    </div>
  </div>
</div>
`;
